### PR TITLE
Functional tests - Add new test 'Change Currency' in FO

### PR DESCRIPTION
**Pull request by sallemiines**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | Add new test 'Change currency' in FO
| Type?             | refacto
| Category?         | TE
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | no
| Related PRs       | no
| How to test?      | no need
| Possible impacts? | no

### CODE REVIEW

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+          await expect(result).to.be.equal(currenciesPage.successfulDeleteMessage);
+        });
+
+        it('should reset filters', async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'resetFilterAfterDelete', baseContext);
+
+          const numberOfCurrenciesAfterReset = await currenciesPage.resetAndGetNumberOfLines(page);
+          await expect(numberOfCurrenciesAfterReset).to.be.equal(numberOfCurrencies);
+        });
+      });
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js:300]:
```suggestion
    
```
You can use the post condition `deleteCurrencyTest` from commonTests

File: tests/UI/pages/FO/currency/index.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  /**
+   * Get the product price value
+   * @param page
+   * @returns {Promise<number>}
+   */
+  async getProductPrice(page) {
+    return this.getPriceFromText(page, this.productPrice);
+  }
+}
+
```
Comment (nesrineabdmouleh) [inline tests/UI/pages/FO/currency/index.js:40]:
```suggestion

```
No need to create new page, you can use all methods on `foBasePage`

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        const createdCurrency = await currenciesPage.getCurrencyFromTable(page, 1);
+        await Promise.all([
+          expect(createdCurrency.name).to.contains(Currencies.toman.name),
+          expect(createdCurrency.symbol).to.contains(Currencies.toman.symbol),
+          expect(createdCurrency.isoCode).to.contains(Currencies.toman.isoCode),
+          expect(createdCurrency.exchangeRate).to.be.above(0),
+          expect(createdCurrency.enabled).to.be.equal(Currencies.toman.enabled),
+        ]);
+      },
+    );
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js:169]:
```suggestion
  
```
You can use the function `createCurrencyTest` from commonTests

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        });
+
+        it('should check the exchange rate value ', async function () {
+          await testContext.addContextItem(this, 'testIdentifier', 'checkExchangeRateValue', baseContext);
+
+          // Check the exchange rate value
+
+          const productWithExchangeRate = await currencyPage.getProductPrice(page);
+          await expect(Math.round(productWithExchangeRate)).to.equal(Math.round(Currencies.euro.exchangeRate * Products.demo_11.finalPrice));
+        });
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js:254]:
```suggestion
       it('should check the product price', async function () {
        await testContext.addContextItem(this, 'testIdentifier', 'checkProductPrice', baseContext);
        
        const productPrice = Math.round(Currencies.toman.exchangeRate * Products.demo_11.finalPrice);

        const productPrice = await currencyPage.getProductPrice(page);
        await expect(productPrice).to.contains(`${Currencies.euro.symbol} ${productPrice}`);
      });
```

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+Add new currency
+Go to FO
+Change the current currency
+Check currency is changed
+Check the product price value is changed
+Check the exchange rate value
+Switch back to the original currency
+Check currency is changed
+Go to Bo
+Delete currency
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js:49]:
```suggestion
Scenario:
 - Go to FO and check that the currencies list is not visible in the header
 - Create new currency
 - Go to FO and switch to the new currency
 - Check that the price of the first product is changed 
 - Switch back to the original currency and check the price of the first product
Post-condition:
- Delete the created currency
```

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+      const isVisible = await homePage.isCurrencyVisible(page);
+      await expect(isVisible).to.be.false;
+    });
+  });
+
+  // Pre-condition: Create currency
+  createCurrencyTest(Currencies.mad, `${baseContext}_preTest_2`);
+
+  describe('Filter by iso code of currency and check the exchange rate value ', async () => {
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js]: Get the exchange rate value or check it ?

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    it('should login in BO', async function () {
+      await testContext.addContextItem(this, 'testIdentifier', 'login', baseContext);
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js]: No need to add a test identifier for loginBO

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+      await testContext.addContextItem(this, 'testIdentifier', 'searchProduct', baseContext);
+
+      await homePage.searchProduct(page, Products.demo_11.name);
+
+      const pageTitle = await searchResultsPage.getPageTitle(page);
+      await expect(pageTitle).to.equal(searchResultsPage.pageTitle);
+    });
+
+    it('should check the product price', async function () {
+      await testContext.addContextItem(this, 'testIdentifier', 'checkProductPrice', baseContext);
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js:152]: Please add an empty line

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        baseContext,
+      );
+
+      // Check currency
+      await homePage.changeCurrency(page, Currencies.euro.isoCode, Currencies.euro.symbol);
+      const shopCurrency = await homePage.getDefaultCurrency(page);
+      await expect(shopCurrency).to.contains(Currencies.euro.isoCode);
+    });
+
+    it('should search product', async function () {
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js]: Maybe no need to search the product another time

File: tests/UI/pages/BO/international/currencies/index.js [hunk 10]
```diff
@@ -153,6 +153,16 @@ class Currencies extends LocalizationBasePage {
     return this.getNumberFromText(page, this.tableColumn(row, 'conversion_rate'));
   }
 
+  /**
+   * Get price from exchange rate value
+   * @param page {Page} Browser tab
+   * @param row
+   * @returns {Promise<number>}
+   */
+  async getPriceFromExchangeRate(page, row) {
```
Comment (nesrineabdmouleh) [inline tests/UI/pages/BO/international/currencies/index.js]: is it get price from exchange rate or get exchange rate value ?

File: tests/UI/pages/FO/FObasePage.js [hunk 11]
```diff
@@ -438,6 +438,16 @@ class FOBasePage extends CommonPage {
...
+
+  /**
+   * Check that currency is visible
+   * @param page {Page} Browser tab
+   * @returns {Promise<boolean>}
+   */
+  async isCurrencyVisible(page) {
+    return this.elementVisible(page, this.currencySelectorDiv, 1000);
+  }
+
```
Comment (nesrineabdmouleh) [inline tests/UI/pages/FO/FObasePage.js]: Please check the lint errors

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    });
+
+    it('should check the product price', async function () {
+      await testContext.addContextItem(this, 'testIdentifier', 'checkProductPrice', baseContext);
+      const newExchangeRateValue = (exchangeRateValue * Products.demo_11.finalPrice).toFixed(Currencies.mad.decimals);
+      const productPrice = await searchResultsPage.getProductPrice(page);
+      await expect(productPrice).to.contains(`${Currencies.mad.symbol}${newExchangeRateValue}`);
+    });
+
+    it('should switch back the the original currency', async function () {
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js]:
```suggestion
    it('should switch back to the default currency', async function () {
```

**Comment by nesrineabdmouleh** [comment]
Please check the doubles steps identifiers

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+    it('should go to \'Currencies\' page', async function () {
+      await testContext.addContextItem(this, 'testIdentifier', 'goToCurrenciesPage', baseContext);
+
+      await localizationPage.goToSubTabCurrencies(page);
+      const pageTitle = await currenciesPage.getPageTitle(page);
+      await expect(pageTitle).to.contains(currenciesPage.pageTitle);
+    });
+
+    it(`should check the currency exchange rate value '${Currencies.mad.isoCode}'`, async function () {
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js]:
```suggestion
    it(`should get the currency exchange rate value '${Currencies.mad.isoCode}'`, async function () {
```

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+      await testContext.addContextItem(this, 'testIdentifier', 'checkExchangeRate', baseContext);
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js]:
```suggestion
      await testContext.addContextItem(this, 'testIdentifier', 'getExchangeRate', baseContext);
```

File: tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+
+    it('should check the product price', async function () {
+      await testContext.addContextItem(this, 'testIdentifier', 'checkProductPrice', baseContext);
+
+      const newExchangeRateValue = (exchangeRateValue * Products.demo_11.finalPrice).toFixed(Currencies.mad.decimals);
+      const productPrice = await searchResultsPage.getProductPrice(page);
+      await expect(productPrice).to.contains(`${Currencies.mad.symbol}${newExchangeRateValue}`);
+    });
+
+    it('should switch back to the original currency', async function () {
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/FO/02_headerAndFooter/03_changeCurrency.js]:
```suggestion
    it('should switch back to the default currency', async function () {
```

**Review by nesrineabdmouleh — DISMISSED** [review]

File: tests/UI/pages/FO/searchResults/index.js [hunk 16]
```diff
@@ -89,6 +90,15 @@ class SearchResults extends FOBasePage {
...
   }
+
+  /**
+   * Get the product price value
+   * @param page
+   * @returns {Promise<string>}
+   */
+  async getProductPrice(page) {
+    return this.getTextContent(page, this.productPrice);
+  }
```
Comment (boubkerbribri) [inline tests/UI/pages/FO/searchResults/index.js]: FYI : if there is no await on the function, you don't need the `async` in function definition. 

```suggestion
  getProductPrice(page) {
    return this.getTextContent(page, this.productPrice);
  }
```

**Review by boubkerbribri — DISMISSED** [review]

**Review by boubkerbribri — DISMISSED** [review]

**Review by boubkerbribri — DISMISSED** [review]
LGTM

**Comment by nesrineabdmouleh** [comment]
Hello @sallemiines,

Please fix conflicting files.

Thank you!

**Comment by boubkerbribri** [comment]
> Hello @sallemiines,
> 
> Please fix conflicting files.
> 
> Thank you!

Done

**Review by boubkerbribri — APPROVED** [review]

**Review by NeOMakinG — APPROVED** [review]

**Pull request merged**